### DISCUSSION TITLE
Feature Request: Ignoring null columns when converting to json

### DISCUSSION THREAD

**Discussion by andrearossiavanade** [duckdb/duckdb#9495] [Ideas]
I have a sparse table which I want to export as jsonlist with a lot of null values in it. 
It would be great to avoid writing null columns when converting the data to json, both with the to_json family of functions or when saving to file with the copy command.
In this way the resulting jsonlist file would be much smaller in size.

Example:

Data I have:
| id | value1 | value2 | value3 |
|----|--------|--------|--------|
| 1  | null   | 1      | null   |
| 2  | null   | null   | 1      |
| 3  | 1      | 1      | null   |

Actual Result:
```json
{"id":1,"value1":null,"value2":1,"value3":null}
{"id":2,"value1":null,"value2":null,"value3":1}
{"id":3,"value1":1,"value2":1,"value3":null}
```

Expected Result:

```json
{"id":1,"value2":1}
{"id":2,"value3":1}
{"id":3,"value1":1,"value2":1}
```

**Comment by cmdlineluser** [comment]
+1

As an existing example, Postgres supports this via `json_strip_nulls`

https://www.postgresql.org/docs/16/functions-json.html

- `json_strip_nulls ( json ) → json`
- `jsonb_strip_nulls ( jsonb ) → jsonb`
> Deletes all object fields that have null values from the given JSON value, recursively. 
> Null values that are not object fields are untouched.

**Comment by eaubin** [comment]
You can shallowly remove null values with the macro below:

```
con.execute("""
CREATE OR REPLACE MACRO remove_null_values(js) AS 
map_from_entries(
    [(key, js->key) FOR key in json_keys(js) IF (js->key) IS NOT NULL]
)::JSON;
""")

con.execute("""
SELECT 
remove_null_values('{"id":1,"value1":null,"value2":1,"value3":null}'::JSON) as j1,
remove_null_values('{"id":2,"value1":null,"value2":null,"value3":1}'::JSON) as j2,
remove_null_values('{"id":3,"value1":1,"value2":1,"value3":null}'::JSON) as j3;
""").fetchall()
# [('{"id":1,"value2":1}',
#   '{"id":2,"value3":1}',
#   '{"id":3,"value1":1,"value2":1}')]
```

**Reply by sarath-mec** [reply]
Getting an error now. Can you please help what is the issue here

**Duck DB Version:** v1.2.1

```
---------------------------------------------------------------------------
BinderException                           Traceback (most recent call last)
Cell In[9], line 9
      1 print(conn.execute("SELECT version() AS version;").fetchall())
      2 conn.execute("""
      3 CREATE OR REPLACE MACRO remove_null_values(js) AS
      4 map_from_entries(
      5     [(key, js->key) FOR key in json_keys(js) IF (js->key) IS NOT NULL]
      6 )::JSON;
      7 """)
----> 9 conn.execute("""
     10 SELECT
     11 remove_null_values('{"id":1,"value1":null,"value2":1,"value3":null}'::JSON) as j1,
     12 remove_null_values('{"id":2,"value1":null,"value2":null,"value3":1}'::JSON) as j2,
     13 remove_null_values('{"id":3,"value1":1,"value2":1,"value3":null}'::JSON) as j3,
     14 
     15 """).fetchall()

BinderException: Binder Error: failed to bind function, either: This scalar function does not support lambdas!
 or: Referenced column "js" not found in FROM clause!
```

**Reply by eaubin** [reply]
Idk @sarath-mec, you should file an issue.

**Reply by eaubin** [reply]
There is some problem with the json_extract operator (->) and the semantics of the json nulls must have changed at some point. Instead use:

```
CREATE OR REPLACE MACRO remove_null_values(js) AS 
map_from_entries(
    [(key, json_extract(js, key)) FOR key in json_keys(js) IF json_extract(js,key) <> ('null'::JSON)]
)::JSON;
```

**Comment by Wocordie** [comment]
Can a similar option be added to the COPY TO json export statements ? Right now if I use the COPY TO export in json format, the json contains values with null. Now my application has to read the exported file and remove the null values afterwards.

**Comment by gcr** [comment]
Touching on the above comment, I've had some luck postprocessing exported JSON through `jq`:

```
D COPY tablename TO 'input.json';
$ jq 'walk(values)' input.json > output.json
```

In jq, the `values` filter returns all non-null inputs unchanged and `walk(•)` applies that filter recursively to each part of the input's structure. So `walk(values)` recurses through input objects and arrays, removing all `null` values.

**Comment by docentYT** [comment]
Any updates on this?
I have to stream query results from the big database in json, and there is no need to send null values.

**Comment by mustafahasankhan** [comment]
Hey folks,

This should be available in next release, the PR is merged!
https://github.com/duckdb/duckdb/pull/21748
